includes/classes/Indexable/Post/Post.php

```diff
@@ -95,14 +97,65 @@ public function query_db( $args ) {
 		 */
 		$args = apply_filters( 'ep_index_posts_args', $args );
 
+		// Implement custom pagination for better performance.
+		if ( ! empty( $args['ep_indexing_advanced_pagination'] ) ) {
+			// Apply an initial start_post_id if needed.
+			$args['ep_indexing_start_post_id'] = ! empty( $args['ep_indexing_start_post_id'] ) ? $args['ep_indexing_start_post_id'] : PHP_INT_MAX;
+			$args['ep_indexing_initial_start_post_id'] = ! empty( $args['ep_indexing_initial_start_post_id'] ) ? $args['ep_indexing_initial_start_post_id'] : PHP_INT_MAX;
+
+			// Get a snapshot of the total found_posts just once (is cached). Uses "initial_start_post_id" to ensure query does not change and stays cached.
+			$total_objects_query = new WP_Query( array_merge( $args, [ 'offset' => 0, 'posts_per_page' => 1, 'no_found_rows' => false, 'ep_indexing_skip_last_part_of_range' => true, 'ep_indexing_start_post_id' => $args['ep_indexing_initial_start_post_id'] ] ) );
+
+			// Force the following query args to ensure things work correctly.
+			unset( $args['offset'] );
+			$args = array_merge( $args, [
+				'suppress_filters' => false,
+				'no_found_rows'    => true,
+				'orderby'          => 'ID',
+				'order'            => 'DESC',
+				'paged'            => 1,
+			] );
+		}
+
 		$query = new WP_Query( $args );
 
+		// If the query failed to return any posts, remove the second part of the range to ensure there truly are no matching posts left.
+		if ( empty( $query->posts ) && ! empty( $args['ep_indexing_advanced_pagination'] ) ) {
```

Comment:
So let's say the site has post IDs 1-1,000 of normal posts. Then nothing-ness until ID 250,000. The query with end ranges will be like `WHERE wp_posts.ID > 1000 AND wp_posts.ID < 51000`. This will return no posts, causing the abstracted logic in `index_helper()` to abort out of the loop because it thinks we're out of posts to index.

Reply:
Oh wait, I think I misunderstood and this comment was about line 129 instead. In that case though, I'm not sure of the best option for this still due to the abstraction layer.

Reply:
Added a second function param of `$exclude_total_objects_count` to dictate whether or not we need to fetch the total object count. Then in Command.php we're now only asking for the total object count once, before the loop begins.

Notably, this is backwards compatible with all the other places that use query_db() and still rely on getting that total objects count.